File: current_session/10.py
"""
10. Regular Expression Matching
Hard

Given an input string (s) and a pattern (p), implement regular expression matching with support for '.' and '*'.

'.' Matches any single character.
'*' Matches zero or more of the preceding element.

The matching should cover the entire input string (not partial).

Note:

    s could be empty and contains only lowercase letters a-z.
    p could be empty and contains only lowercase letters a-z, and characters like . or *.

"""
class Solution(object):
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """
        # Implement DP
        # Define P[i][j] = True if s[:i] == p[:j]
        m, n = len(s), len(p)
        dp = [[False for _ in xrange(n+1)] for _ in xrange(m+1)]    # carefully arrange size
        dp[0][0] = True
        #print "inputs: ", s, p, dp
        for i in xrange(m+1):           # including m & n
            for j in xrange(1, n+1):    # arrange j in inner loop => increase p's index one by one
                #print "Now at:", i, j, s[:i], p[:j],
                if p[j-1] == '*':       # latest element
                    dp[i][j] = dp[i][j-2] or (i>0 and (s[i-1] == p[j-2] or p[j-2] == '.') and dp[i-1][j])
                else:
                    dp[i][j] = i > 0 and dp[i-1][j-1] and (s[i-1] == p[j-1] or p[j-1] == '.')
                #print dp[i][j], dp
        
        return dp[-1][-1]

print Solution().isMatch("aa", "a")
print Solution().isMatch("aa", "a*")
print Solution().isMatch("ab", ".*")
print Solution().isMatch("aab", "c*a*b")
print Solution().isMatch("mississippi", "mis*is*p*.")

File: current_session/c++/10.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  bool isMatch(string s, string p) {
    // use DP
    // dp[i][j] is the solution of s[:i] + p[:j]
    // "" "*"
    // "aa" "a"
    // "aa" "a*"
    // "aa" "a."
    // "aa" "a."
    // "ab" "aa*"
    // "aab" "a*bc*"
    // 
    // dp[i][j] = (i-1 < len(s) && j-1 < len(j) && s[i-1] == p[j-1] && dp[i-2][j-2])   if p[j] not in "*."
    //          = (dp[i][j-2]) or (dp[i-1][j-1] and p[j-1]==s[i-1]) "*"
    //          = (dp[i-1][j-1])
    int m = s.size(), n = p.size();
    vector<vector<bool> > dp(m+1, vector<bool>(n+1, false));
    dp[0][0] = true;
    for (int i=0; i<m+1; i++) {
      for (int j=1; j<n+1; j++) {
        if (p[j-1] == '*') {
          dp[i][j] = (dp[i][j-2]) || (i && dp[i-1][j] && (s[i-1] == p[j-2] || p[j-2] == '.'));
        } else if (p[i-1] == '.') {
          dp[i][j] = i && dp[i-1][j-1];
        } else {
          dp[i][j] = i && dp[i-1][j-1] && s[i-1] == p[j-1];
        }
      }
    }
    return dp[m][n];
  }
};
int main() {
  Solution sol = Solution();
  // "" "s*"
  // "aa" "a"
  // "aa" "a*"
  // "aa" "a."
  // "aa" "aa"
  // "ab" "aa*"
  // "aab" "a*bc*"
  vector<string> s{ "","aa","aa","aa","aa","ab","aab" };
  vector<string> p{ "s*","a","a*","a.","aa","aa*","a*bc*" };
  // bool res = sol.isMatch("aa", "a*");
  for (int i=0; i<s.size(); i++) {
    cout<<s[i]<<"   "<<p[i];
    printf(" res : %s\n", sol.isMatch(s[i], p[i]) ? "true" : "false");
  }
  
  return 0;
}